airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py

```diff
@@ -324,19 +346,42 @@ def read_records(
             stream_slice=_slice,
             records_schema=records_schema,
         )
-        for stream_data in self._read_pages(record_generator, self.state, _slice):
-            current_record = self._extract_record(stream_data, _slice)
-            if self.cursor and current_record:
-                self.cursor.observe(_slice, current_record)
 
-            # Latest record read, not necessarily within slice boundaries.
-            # TODO Remove once all custom components implement `observe` method.
-            # https://github.com/airbytehq/airbyte-internal-issues/issues/6955
-            most_recent_record_from_slice = self._get_most_recent_record(most_recent_record_from_slice, current_record, _slice)
-            yield stream_data
+        if self.cursor and isinstance(self.cursor, ResumableFullRefreshCursor):
+            stream_state = self.state
 
-        if self.cursor:
-            self.cursor.close_slice(_slice, most_recent_record_from_slice)
+            # Before syncing the RFR stream, we check if the job's prior attempt was successful and don't need to fetch more records
+            # The platform deletes stream state for full refresh streams before starting a new job, so we don't need to worry about
+            # this value existing for the initial attempt
+            if stream_state.get(FULL_REFRESH_SYNC_COMPLETE_KEY):
```

Comment:
thinking out loud: the fact that there are essentially two different code paths that are mutually exclusive for a given stream makes me think there's an opportunity to introduce a new abstraction where one implementation would be RFR and the other the standard `read_records`. 

I'd love to simplify this, but I'm not sure what the right abstraction is yet

Reply:
Yeah I would agree with that. Its kind of our way of not needing to re-implement `read_records()` like we do for Python sources. It's not immediately clear at what level the abstraction needs to sit at yet. But I can't imagine pulling this out into its own abstraction later will be much of a lift.